File: codeforces/1535d_cpp/1535d_cpp/1535d.cpp
#include <stdio.h>

int K;
int Q;
char S[300000];
int CNTTREE[300000];
int LtS[300000];
int StL[300000];
int LIMIT;

void solve(int p, char c) {
	int si = p - 1;
	S[si] = c;

	int li = StL[si];
	while (li > 0) {
		si = LtS[li];
		if (li >= (LIMIT >> 1)) {
			CNTTREE[li] = (S[si] == '?') ? 2 : 1;
		}
		else {
			if (S[si] == '0') {
				CNTTREE[li] = CNTTREE[li * 2];
			}
			else if (S[si] == '1') {
				CNTTREE[li] = CNTTREE[li * 2 + 1];
			}
			else {
				CNTTREE[li] = CNTTREE[li * 2] + CNTTREE[li * 2 + 1];
			}
		}
		li >>= 1;
	}

	printf("%d\n", CNTTREE[1]);
}

int main() {
	scanf("%d\n", &K);
	scanf("%s\n", S);
	scanf("%d\n", &Q);

	LIMIT = 1 << K;

	int lv_beg = 1;
	int lv_end = 2;

	while (lv_beg < LIMIT) {
		int s_beg = LIMIT - lv_end;
		int s_end = LIMIT - lv_beg;

		for (int i = 0; i < lv_end - lv_beg; ++i) {
			int li = lv_beg + i;
			int si = s_beg + i;
			LtS[li] = si;
			StL[si] = li;
		}

		lv_beg = lv_end;
		lv_end <<= 1;
	}

	int leaf_start = LIMIT >> 1;
	for (int li = leaf_start; li < LIMIT; ++li) {
		int si = LtS[li];
		CNTTREE[li] = (S[si] == '?') ? 2 : 1;
	}

	int curlv_start = leaf_start >> 1;
	int curlv_end = leaf_start;

	while (curlv_start > 0) {
		for (int li = curlv_start; li < curlv_end; ++li) {
			int si = LtS[li];
			if (S[si] == '0') {
				CNTTREE[li] = CNTTREE[li * 2];
			}
			else if (S[si] == '1') {
				CNTTREE[li] = CNTTREE[li * 2 + 1];
			}
			else {
				CNTTREE[li] = CNTTREE[li * 2] + CNTTREE[li * 2 + 1];
			}
		}
		curlv_end = curlv_start;
		curlv_start >>= 1;
	}

	int p;
	char c;
	for (int q = 0; q < Q; ++q) {
		scanf("%d %c\n", &p, &c);
		solve(p, c);
	}
}

File: codeforces/problemset/1535/1535d.py
K,S,Q = None,None,None
LIMIT = None
CNTTREE = None
LtS = None
StL = None

def solve(p,c):
	global S,CNTTREE

	si = p-1
	S[si] = c

	li = StL[si]
	while li > 0:
		si = LtS[li]
		if li >= (LIMIT >> 1):
			CNTTREE[li] = 2 if S[si] == '?' else 1
		else:
			if S[si] == '0':
				CNTTREE[li] = CNTTREE[li*2]
			elif S[si] == '1':
				CNTTREE[li] = CNTTREE[li*2+1]
			else:
				CNTTREE[li] = CNTTREE[li*2] + CNTTREE[li*2+1]
		li >>= 1

	# print(S,CNTTREE)
	print(CNTTREE[1])


if __name__ == '__main__':
	K = int(input())
	S = list(input())
	Q = int(input())

	LIMIT = 2**K
	CNTTREE = [0]*LIMIT

	LtS = [None]*LIMIT
	StL = [None]*LIMIT

	# compute idx transitions
	lv_beg = 1
	lv_end = 2
	li = 1
	while lv_beg < LIMIT:
		li = lv_beg
		s_beg = LIMIT-lv_end
		s_end = LIMIT-lv_beg

		for i in range(lv_end-lv_beg):
			li = lv_beg+i
			si = s_beg+i
			LtS[li] = si
			StL[si] = li

		lv_beg, lv_end = lv_end, lv_end << 1

	# print(f'LtS: {LtS}')
	# print(f'StL: {StL}')

	leaf_start = LIMIT >> 1
	for li in range(leaf_start,LIMIT):
		si = LtS[li]
		CNTTREE[li] = 2 if S[si] == '?' else 1

	curlv_start = leaf_start >> 1
	curlv_end = leaf_start
	while curlv_start > 0:
		for li in range(curlv_start, curlv_end):
			si = LtS[li]
			if S[si] == '0':
				CNTTREE[li] = CNTTREE[li*2]
			elif S[si] == '1':
				CNTTREE[li] = CNTTREE[li*2+1]
			else:
				CNTTREE[li] = CNTTREE[li*2]+CNTTREE[li*2+1]
		curlv_start, curlv_end = curlv_start >> 1, curlv_start

	# print(S,CNTTREE)

	for q in range(Q):
		p,c = input().split()
		p = int(p)
		solve(p,c)